tests/e2e/etcd_mix_versions_test.go

```diff
@@ -0,0 +1,213 @@
+	"time"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.etcd.io/etcd/client/pkg/v3/fileutil"
+	"go.etcd.io/etcd/tests/v3/framework/e2e"
+)
+
+// TestMixVersionsSnapshotByAddingMember tests the mix version send snapshots by adding member
+func TestMixVersionsSnapshotByAddingMember(t *testing.T) {
+	cases := []struct {
+		name               string
+		clusterVersion     e2e.ClusterVersion
+		newInstanceVersion e2e.ClusterVersion
+	}{
+		// TODO: uncomment after v3.4.32 release.
+		// After v3.4.32, etcd can support adding a new member of 3.4 into
+		// a cluster with 3.5 if the 3.4 member is started with '--next-cluster-version-compatible'.
+		/*{
+			name:              "etcd instance with last version receives snapshot from the leader with current version",
+			clusterVersion:    e2e.CurrentVersion,
+			newInstaceVersion: e2e.LastVersion,
+		},*/
+		{
+			name:               "etcd instance with current version receives snapshot from the leader with last version",
+			clusterVersion:     e2e.LastVersion,
+			newInstanceVersion: e2e.CurrentVersion,
+		},
+	}
+
+	for _, tc := range cases {
+		t.Run(tc.name, func(t *testing.T) {
+			mixVersionsSnapshotTestByAddingMember(t, tc.clusterVersion, tc.newInstanceVersion)
+		})
+	}
+}
+
+func mixVersionsSnapshotTestByAddingMember(t *testing.T, clusterVersion, newInstanceVersion e2e.ClusterVersion) {
+	if !fileutil.Exist(e2e.BinPathLastRelease) {
+		t.Skipf("%q does not exist", e2e.BinPathLastRelease)
+	}
+
+	e2e.BeforeTest(t)
+
+	// Create an etcd cluster with 1 member
+	copiedCfg := e2e.NewConfigNoTLS()
+	copiedCfg.ClusterSize = 1
+	copiedCfg.Version = clusterVersion
+	copiedCfg.SnapshotCount = 10
+	copiedCfg.BasePeerScheme = "unix" // to avoid port conflict
+
+	epc, err := e2e.NewEtcdProcessCluster(t, copiedCfg)
+
+	require.NoError(t, err, "failed to start etcd cluster: %v", err)
+	defer func() {
+		derr := epc.Close()
+		require.NoError(t, derr, "failed to close etcd cluster: %v", derr)
+	}()
+
+	// Write more than SnapshotCount entries to trigger at least a snapshot.
+	t.Log("Writing 20 keys to the cluster")
+
+	etcdctl := e2e.NewEtcdctl(epc.Procs[0].EndpointsV3(), e2e.ClientNonTLS, false, false)
+	for i := 0; i < 20; i++ {
+		key := fmt.Sprintf("key-%d", i)
+		value := fmt.Sprintf("value-%d", i)
+		err = etcdctl.Put(key, value)
+		require.NoError(t, err, "failed to put %q, error: %v", key, err)
+	}
+
+	// start a new etcd instance, which will receive a snapshot from the leader.
+	newCfg := *epc.Cfg
+	newCfg.Version = newInstanceVersion
+	t.Log("Starting a new etcd instance")
+	_, err = epc.StartNewProc(&newCfg, t)
+	require.NoError(t, err, "failed to start the new etcd instance: %v", err)
+	defer epc.Close()
+
+	// verify all nodes have exact same revision and hash
+	t.Log("Verify all nodes have exact same revision and hash")
+	assert.Eventually(t, func() bool {
+		hashKvs, err := e2e.NewEtcdctl(epc.EndpointsV3(), e2e.ClientNonTLS, false, false).HashKV(0)
```

Comment:
I don't think this validation is very useful. I expect this check to pass even without fixes to migrating the snapshot. 

`HashKV` only checks the contents of Key bucket, ignoring all the other metadata. As the layout of key bucket didn't change at between v3.5 and v3.4, this doesn't really check if data was properly migrated. 

Options:
* Use `Hash` that returns hash of whole etcd storage, but skips some keys from metadata. Not sure it will work across etcd version.
* Check other property to verify that data was properly migrated.

Reply:
Added https://github.com/etcd-io/etcd/issues/17752 as followup. But I don't think that should necessarily block this PR.

Reply:
SG